src/qml_benchmarks/models/base.py

```diff
@@ -0,0 +1,299 @@
+        return configs
+
+    def langevin_sample(self, params, x_init, n_samples, key):
+        pass
+
+    def sample(self, num_samples, num_mcmc_steps=1000):
+        """
+        sample configurations starting from a random configuration.
+        """
+        if self.params_ is None:
+            raise ValueError(
+                "Model not initialized. Call model.initialize first with"
+                "example data sample."
+            )
+        keys = jax.random.split(self.generate_key(), num_samples)
+
+        # we do not take the gradient wrt the sampling, so decouple the param dict here
+        params_copy = copy.deepcopy(self.params_)
+
+        for key in params_copy.keys():
+            params_copy[key] = jax.lax.stop_gradient(params_copy[key])
+
+        x_init = jnp.array(
+            jax.random.bernoulli(
+                self.generate_key(), p=0.5, shape=(num_samples, self.dim)
+            ),
+            dtype=int,
+        )
+        configs = self.batched_mcmc_sample(params_copy, x_init, num_mcmc_steps, keys)
+        x1 = configs[:, -1]
+        return x1
+
+    def contrastive_divergence_loss(self, params, X, y, key):
+        """
+        Contrastive divergence loss function.
+        Args:
+            X (array): batch of training examples
+            y (array): not used; should be set to None when training
+            key: jax PRNG key
+        """
+        keys = jax.random.split(key, X.shape[0])
+
+        # we do not take the gradient wrt the sampling, so decouple the param dict here
+        params_copy = copy.deepcopy(params)
+        for key in params_copy.keys():
+            params_copy[key] = jax.lax.stop_gradient(params_copy[key])
+
+        configs = self.batched_mcmc_sample(params_copy, X, self.cdiv_steps, keys)
+        x0 = configs[:, 0]
+        x1 = configs[:, -1]
+
+        # taking the gradient of this loss is equivalent to the CD-k update
+        loss = self.energy(params, x0) - self.energy(params, x1)
+
+        return jnp.mean(loss)
+
+    def fit(self, X: jnp.array, y: any = None) -> None:
+        """
+        Fit the parameters and update self.params_.
+        """
+        self.initialize(X)
+        c_div_loss = (
+            jax.jit(self.contrastive_divergence_loss)
+            if self.jit
+            else self.contrastive_divergence_loss
+        )
+
+        self.params_ = train(
+            self,
+            c_div_loss,
+            optax.adam,
+            X,
+            None,
+            self.generate_key,
+            convergence_interval=self.convergence_interval,
+        )
+
+    def score(self, X, y=None):
+        """Score the model on the given data.
+
+        Higher is better.
+        """
+        if self.params_ is None:
+            self.initialize(X.shape[1])
+
+        c_div_loss = (
+            jax.jit(self.contrastive_divergence_loss)
+            if self.jit
+            else self.contrastive_divergence_loss
+        )
+
```

Comment:
have been thinking a bit about this. I think one option for the score that makes sense is the pseudo likelihood, which is used in the RBM I believe. 

Once could also use the maximum mean discrepancy. I'll probably add this to the package at some point. 

We can leave like this for now but we should come back to this.